File: backend/app/routes/root_cause.py
from fastapi import APIRouter, Depends
from pydantic import BaseModel

from app.models.user import User
from app.security.dependencies import get_current_user
from app.services import root_cause_service

router = APIRouter(prefix="/api/testing/root-cause", tags=["root-cause"])


class AnalyzeRequest(BaseModel):
    error_message: str
    stack_trace: str = ""
    test_case_title: str = ""


class AnalyzeResultOut(BaseModel):
    id: str
    root_cause: str
    confidence: float
    recommendation: str
    affected_component: str
    likely_commit_sha: str | None
    likely_commit_message: str | None
    source: str
    git_correlation_available: bool


class HistoryEntryOut(BaseModel):
    id: str
    test_case_title: str
    root_cause: str
    confidence: float
    likely_commit_sha: str | None
    source: str


@router.post("/{project_id}/analyze", response_model=AnalyzeResultOut)
async def analyze(project_id: str, payload: AnalyzeRequest, user: User = Depends(get_current_user)):
    result = await root_cause_service.analyze_with_correlation(
        user.organization_id, project_id, payload.error_message, payload.stack_trace, payload.test_case_title
    )
    return AnalyzeResultOut(**result)


@router.get("/{project_id}/history", response_model=list[HistoryEntryOut])
async def history(project_id: str, user: User = Depends(get_current_user)):
    entries = await root_cause_service.list_history(user.organization_id, project_id)
    return [HistoryEntryOut(**e.model_dump()) for e in entries]

File: frontend/src/services/api/rootCause.ts
import { apiClient } from "./client";

export interface AnalyzeResultOut {
  id: string;
  root_cause: string;
  confidence: number;
  recommendation: string;
  affected_component: string;
  likely_commit_sha: string | null;
  likely_commit_message: string | null;
  source: string;
  git_correlation_available: boolean;
}

export interface HistoryEntryOut {
  id: string;
  test_case_title: string;
  root_cause: string;
  confidence: number;
  likely_commit_sha: string | null;
  source: string;
}

export async function analyzeRootCause(
  projectId: string, errorMessage: string, stackTrace: string, testCaseTitle: string
): Promise<AnalyzeResultOut> {
  const { data } = await apiClient.post<AnalyzeResultOut>(`/api/testing/root-cause/${projectId}/analyze`, {
    error_message: errorMessage, stack_trace: stackTrace, test_case_title: testCaseTitle,
  });
  return data;
}

export async function listRootCauseHistory(projectId: string): Promise<HistoryEntryOut[]> {
  const { data } = await apiClient.get<HistoryEntryOut[]>(`/api/testing/root-cause/${projectId}/history`);
  return data;
}
